deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/FancyBlockingQueue.java

```diff
@@ -0,0 +1,288 @@
+package org.deeplearning4j.optimize.solvers.accumulation;
+
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.deeplearning4j.exception.DL4JInvalidInputException;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.LockSupport;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * This BlockingQueue implementation is suited only for symmetric gradients updates, and should NOT be used anywhere else.
+ *
+ * Basic idea: all worker threads requesting via poll()/take() method will be advancing only once all consumers get the same element from Queue.
+ * So, multiple consumers are guaranteed to be consuming the same elements in the same order served by this queue.
+ *
+ * @author raver119@gmail.com
+ */
+@Slf4j
+public class FancyBlockingQueue<E> implements BlockingQueue<E>, Registerable {
```

Comment:
Plus there's one more thing. One of possible further improvements there, will be local averaging of updates before sharing them over network. To make that possible - updates should be symmetrical.